site/app/libraries/database/DatabaseQueries.php

```diff
@@ -7175,6 +7176,160 @@ public function getGradeableAccessTeam(
+            return new \EmptyIterator();
+        }
+        if ($paths === null) {
+            $paths = [];
+        }
+
+        // Generate the selector statement
+        $selector = '';
+        if (count($paths) > 0) {
+            $place_holders = implode(',', array_fill(0, count($paths), '?'));
+            $selector = "WHERE cm.url IN ($place_holders)";
+        }
+
+        $query = "SELECT 
+                    cm.path,
+                    cm.type,
+                    cm.release_date,
+                    cm.hidden_from_students,
+                    cm.priority,
+                    cm.section_lock,
+                    json_agg(cms.section_id) AS sections
+                        FROM course_materials AS cm
+                        LEFT JOIN course_materials_sections AS cms ON cm.path = cms.course_material_id
+                    $selector GROUP BY cm.path;";
+        $course_material_constructor = function ($row) {
+            $row['sections'] = json_decode($row['sections']);
+            return new CourseMaterial($this->core, $row);
+        };
+        return $this->course_db->queryIterator(
+            $query,
+            $paths,
+            $course_material_constructor
+        );
+    }
+
+    public function getCourseMaterial($path): CourseMaterial {
+        foreach ($this->getCourseMaterials([$path]) as $course_material) {
+            return $course_material;
+        }
+        throw new \InvalidArgumentException('Course material does not exist');
+    }
+
+    public function createCourseMaterial(CourseMaterial $course_material): bool {
+        try {
+            $params = [
+                $course_material->getType(),
+                $course_material->getPath(),
+                $course_material->getReleaseDate(),
+                $course_material->getHiddenFromStudents(),
+                $course_material->getPriority(),
+                $course_material->getSectionLock()
+            ];
+            $this->course_db->query(
+                "
+            INSERT INTO course_materials(
+              type,
+              path,
+              release_date,
+              hidden_from_students,
+              priority,
+              section_lock)
+            VALUES (?, ?, ?, ?, ?, ?)",
+                $params
+            );
+            if ($course_material->getSectionLock()) {
+                foreach ($course_material->getSections() as $section) {
+                    $params = [
+                        $course_material->getPath(),
+                        $section
+                    ];
+                    $this->course_db->query(
+                        "
+                INSERT INTO course_materials_sections(
+                  course_material_id,
+                  section_id)
+                VALUES (?, ?)",
+                        $params
+                    );
+                }
+            }
+            return true;
+        }
+        catch (DatabaseException $e) {
```

Comment:
why are you catching this exception?

Reply:
This is to answer this one and all the ones below. I did this so the controller could tell if the query was successful. The main reason I did this was for testing to mock those functions and return a true or false. Can you look at the CourseMaterialsControllerTester.php and see if there is a better way I can check and validate the whole DatabaseQueries functions?

Reply:
You could use:

https://github.com/Submitty/Submitty/blob/d76e545d5ec67e466265ba67d05d085607f60ae0/site/app/libraries/database/AbstractDatabase.php#L316-L327

to determine the affected number of rows? Depending on what you're trying to catch here (insert duplicate path?), I would suggest adding a check within the controller to protect against that condition  there, and the DB query operates expecting it to always succeed.

Catching the `DatabaseException` will catch both whatever normal condition you're expecting, but also if the query gets malformed, database dies unexpectedly, etc, which we definitely do not want to catch and is something we want to log.

Reply:
This is how I currently verify if the argument is valid.

https://github.com/Submitty/Submitty/blob/7617456fe76d59e9a7187d76a30385e5d801c014/site/tests/app/controllers/course/CourseMaterialsControllerTester.php#L101-L120

This would make sure the information being passed into the function is correct which is what I want to do be doing. How can I switch this around so I don't have to use the true/false values with the catches? I am not necessarily looking to see the number of rows changed but rather what data went in. Or am I doing testing incorrectly?